app/services/setting_registry.rb

```diff
@@ -0,0 +1,107 @@
+class SettingRegistry
+  include Singleton
+
+  def initialize
+    @settings = {}
+  end
+
+  def ready?
+    @settings.any?
+  end
+
+  def logger
+    Foreman::Logging.logger('app')
+  end
+
+  def find(name)
+    logger.warn("Setting is not initialized yet, requested value for #{name} will be always nil") unless ready?
+    @settings[name.to_s]
+  end
+
+  # Returns a setting effective value
+  # Call as `Foreman.settings[<name>]`
+  #
+  #   Foreman.settings['default_locale'] => nil.
+  #   Foreman.settings[:default_locale] => nil.
+  #
+  def [](name)
+    definition = find(name)
+    unless definition
+      logger.warn("Setting #{name} has no definition, please define it before using") if ready?
+      return
+    end
+    definition.value
+  end
+
+  def []=(name, value)
+    definition = find(name)
+    raise ::Foreman::Exception.new(N_("Setting definition for '%s' not found, can not set"), name) unless definition
+    db_record = _find_or_new_db_record(name)
+    db_record.value = value
+    db_record.save!
+    definition.value = db_record.value
+  end
+
+  def load
+    # add() all setting definitions
+    Setting.descendants.each do |cat_cls|
+      if cat_cls.default_settings.empty?
+        # Setting category uses really old way of doing things
+        _load_category_from_db(cat_cls)
+      else
+        cat_cls.default_settings.each { |s| _add(s[:name], s.except(:name).merge(category: cat_cls.name)) }
+      end
+    end
+
+    # load all db values
+    load_values
+
+    # if create_missing create missing values in DB
+    # not needed now as old load mechanism takes care of that
+  end
+
+  def load_values
+    loaded_names = []
+    Setting.unscoped.all.each do |set|
+      set_definition = find(set.name)
+      logger.debug("Setting #{set.name} has no definition, clean up your database") and next unless set_definition
+      loaded_names << set.name
+      set_definition.updated_at = set.updated_at
+      set_definition.value = set.value
+    end
+    # load nil to set value to default
+    @settings.except(*loaded_names).each do |name, set_definition|
+      set_definition.updated_at = nil
+      set_definition.value = set_definition.default
+    end
+  end
+
+  def _add(name, category:, default:, description:, full_name: nil, value: nil, encrypted: false)
```

Comment:
Should there be a `private` before this:
```suggestion
  private

  def _add(name, category:, default:, description:, full_name: nil, value: nil, encrypted: false)
```

Reply:
we are using it in tests, I prefer to have `_add`, but keep it public, so it is possible to use, if necessary, but known it is not public API

Reply:
In tests you can use `.send(:add)` to call private methods, right? I think I'm fine either way. I come from a Python world where there are no real private methods, just underscore convention. I was mostly thinking about it because tools that generate API docs may not respect these "private" methods.

Reply:
yeah, you can, though I like this more, I believe it is more readable, I didn't want to refuse access to this method, just hint it is not public API.
Should I add :nodoc: ?
It would be good idea, but we don't have documentation tool to target this comment to, so I'm not sure it makes sense.

Reply:
Let's leave it for a documentation update then.